Extract out "Enter Nickname" page and rebuild the "Create" page.

zpChris opened the pull request:
### List of Changes:
- Extract out the "Enter Nickname" page as a core component, which both `Create.tsx` and `Join.tsx` use.
- Rebuild the `Create.tsx` page, which redirects to the "Waiting Room" state of the `Join.tsx` page after the nickname is entered.
- Construct logic flow for handling errors, redirects, and state changes between the Create and Join pages.

### Notes:
- This PR, in my opinion, is fairly complicated -- simplification is something I could use a lot of help with in these reviews. Here is an explanation of how the Join Room and Create Room pages now work:

**Join Room**
1. The join room page has two states: (1) The "Enter Nickname" page, and (2) the "Waiting Room" page. These are both handled by the `pageState` React hook.
2. When you initially visit `/game/join`, you are on page 1. There are no `users`, the socket is not connected, and the nickname is initially `''`. I set the content to be the `<EnterNicknamePage ... />`, where I pass in the type of Enter Nickname page (`enterNicknamePageType`) and the function of what to do after the user enters their nickname (`enterNicknameAction`).
3. When the user enters their nickname, they do the following (catching errors all the while):
  a. Connect to the socket at the predefined `SOCKET_ENDPOINT`.
  b. Subscribe to future messages on the socket at `SUBSCRIBE_URL`.
  c. Add the user to the room by sending a message to the socket, with their `nickname` as a parameter.
  d. On the join page, set the `nickname`, `socketConnected` status, and finally update the `pageState` to 2 (Waiting Room).
4. After this function runs, the page will now be on `pageState === 2`, and thus display the waiting room page. This receives any socket updates to the `users` variable as well.

**Create Room**
1. The create room page only has one state -- the `<EnterNicknamePage ... />`.
2. When the user enters their nickname, they do the following (catching errors all the while):
  a. Create the `roomParams` for the `createRoom()` function.
  b. Create the Room in the backend by calling `createRoom(roomParams)`.
  c. Redirect to the "Waiting Room" page (`pageState == 2` on the Join Room page).
3. After this function runs, the page will now be on `pageState === 2` on the Join Room page (thus, the "Waiting Room"). However, the socket is not yet connected.* Therefore, I have the following `if` statement, which verifies that if the socket is not connected but the user is on the "Waiting Room" page, to connect them to the socket.
```
if (!socketConnected && pageState === 2 && nickname) {
  addUserToWaitingRoom(SOCKET_ENDPOINT, SUBSCRIBE_URL, nickname);
}
```
4. The above function connects, subscribes, and adds the user to the page. You are now fully connected on the "Waiting Room" page.

Clearly, this flow of logic is complicated. I tried to elucidate how it works above and in the comments, but _please_ let me know what improvements you may be able to think of!

\* I cannot currently connect the socket in the `Create.tsx` file because the `subscribeCallback` function calls the `setUsers` React hook in the `Join.tsx` file. This is what controls what users are present in the "Waiting Room". Unless I pass this as a parameter to the `Create.tsx` file somehow, the `Create.tsx` cannot properly complete the `subscribeCallback` function.

### Screencast and Images:
Not many changes were made to the UI, so see the screencasts for the user flow.

[Screencast of the User Flow Between Create and Join Pages](https://www.loom.com/share/581763b307054fc69e1967feb914d7b5)

zpChris commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Rebuild create room page, modularize "enter nickname" page](https://trello.com/c/Bthvystn/123-rebuild-create-room-page-modularize-enter-nickname-page)

alankbi reviewed (changes requested):
Mainly smaller nits/suggestions, and then a bigger suggestion which is to move the waiting room stuff to a separate file/room (as I think that makes more sense both UX wise code-wise). However, we already have a card for that in the backlog, so if you decide we should do that in a separate PR, then I think this overall looks good for approval! 

Also, I caught two potential bugs:

1. When you create a room, navigate back to home, then try to join a room, you get the socket already connected error
2. When a creator of a room has their name removed from the waiting page, it shows up again whenever you refresh

I think these are okay though, since extracting into a new waiting room page plus implementing the leave room functionality will naturally solve both of these in time - just wanted to point them out for reference.

zpChris commented on frontend/src/components/core/Button.tsx:
```diff
@@ -29,15 +28,3 @@ export const PrimaryButton = styled(DefaultButton)<any>`
   min-height: 70px;
 `;
-
-export const TextButton = styled.button<ThemeType>`
-  background: none;
-  border: none;
-  color: ${({ theme }) => theme.colors.gray};
-  font-size: ${({ theme }) => theme.fontSize.default};
-  cursor: pointer;
-  
-  &:focus {
-    outline: none;
-  }
-`;
```
Sure, no problem 👍

zpChris commented on frontend/src/views/Join.tsx:
```diff
@@ -1,41 +1,54 @@
-  // Declare nickname state variable.
-  const [nickname, setNickname] = useState('');
+  // Grab initial state variables if navigated from the create page.
+  const location = useLocation<JoinGamePageProps>();
+  const joinGamePageProps: JoinGamePageProps = {};
+  if (location && location.state) {
+    joinGamePageProps.initialPageState = location.state.initialPageState;
+    joinGamePageProps.initialNickname = location.state.initialNickname;
```
> We do have another story for that though so if you prefer not to include it in this PR, that's totally fine

Ok, that sounds good -- in that case, I'll do this in the next PR that I work on.

zpChris commented on frontend/src/views/Create.tsx:
```diff
@@ -0,0 +1,42 @@
+  const history = useHistory();
+
+  // Creates a room with the user as the host, and joins that same waiting room.
+  const createJoinWaitingRoom = (nickname: string) => new Promise<undefined>((resolve, reject) => {
```
The reason why I make Promises for both of these is because I need to know when to call `setLoading(false)`. If this wasn't a promise, and instead simply a method, then I would either just never call `setLoading(false)` (which seems strange), or include it right after, which would then immediately disable the loading element as it would not wait for this method to finish. I could add a callback to call `setLoading(false)`, but at that point since I also have an error, I figured I might as well make an undefined Promise.

zpChris commented on frontend/src/views/Join.tsx:
```diff
@@ -52,25 +65,43 @@ function JoinGamePage() {
+   * This method returns a Promise which is used to trigger setLoading
+   * and setError on the EnterNickname page following this function.
    */
   const addUserToWaitingRoom = (socketEndpoint: string,
```
A couple things IMO:
1. The `setError(err.message)` function should be for the `EnterNickname.tsx` page, _not_ the `Join.tsx` page. Thus, I'd have to pass in this set function as a parameter.
2. See my [other comment](https://github.com/rocketden/main/pull/36#discussion_r492099356).

zpChris commented on frontend/src/components/core/EnterNickname.tsx:
```diff
@@ -0,0 +1,102 @@
+import { isValidNickname } from '../../api/Socket';
+
+// The type of enter nickname page (create or join).
+export const ENTER_NICKNAME_PAGE = Object.freeze({
+  CREATE: 'create',
+  JOIN: 'join',
+});
```
I made this initially because I thought I could use this for if statements if there were different design changes on create vs. join pages; however, I think you're right that this would just be better as a parameter.

Fixed!

zpChris commented:
> Mainly smaller nits/suggestions, and then a bigger suggestion which is to move the waiting room stuff to a separate file/room (as I think that makes more sense both UX wise code-wise). However, we already have a card for that in the backlog, so if you decide we should do that in a separate PR, then I think this overall looks good for approval!
> 
> Also, I caught two potential bugs:
> 
> 1. When you create a room, navigate back to home, then try to join a room, you get the socket already connected error
> 2. When a creator of a room has their name removed from the waiting page, it shows up again whenever you refresh
> 
> I think these are okay though, since extracting into a new waiting room page plus implementing the leave room functionality will naturally solve both of these in time - just wanted to point them out for reference.

Thanks for the nits -- I think I'll do that suggested nit in a separate PR (or maybe you or @helenli8181 could actually do it, given that maybe we should start to work on each other's code to introduce some more familiarity). I have noticed (1) before, and (2) is actually good -- we'll want the person to reconnect if they leave and join again, so I think that design will actually be helpful in the future. I also agree that those errors will be fixed as we build out the leave room functionality.

I addressed most of your nits, some of them I responded to instead (mostly the `Promise<undefined>` ones I thought may need to be left in there).

alankbi reviewed (approved):
Thanks for addressing my comments! Approved

helenli8181 commented on frontend/src/components/core/EnterNickname.tsx:
```diff
@@ -0,0 +1,92 @@
+      />
+      <LargeInputButton
+        onClick={() => {
+          setLoading(true);
+          enterNicknameAction(nickname).then(() => {
+            setLoading(false);
+          }).catch((err: ErrorResponse) => {
+            setLoading(false);
+            setError(err.message);
+          });
```
Would it be better to extract these lines into a function to avoid repeating them? I think that's what you did in `addUsersToWaitingRoom()`.

zpChris replied:
I think that makes sense, I'll update this and makes sure it works -- thanks!

zpChris replied:
Updated 👍

helenli8181 commented on frontend/src/views/Join.tsx:
```diff
@@ -1,41 +1,54 @@
+   * Nickname that is populated if the join page is on the waiting room stage.
+   * Set error if no nickname is passed in despite the waiting room stage.
+   */
+  const [nickname, setNickname] = useState(joinGamePageProps.initialNickname || '');
+  if (pageState === 2 && !nickname && !error) {
+    setError('No nickname was provided for the user in the waiting room.');
+  }
```
I'm kind of confused, is it possible for a user to already be in the waiting room but not have a nickname?

zpChris replied:
It shouldn't be possible -- I primarily included it to prevent issues from the developer's perspective. This could also be used to guard against malicious users, who somehow reach the second page state without a nickname.

helenli8181 reviewed (approved):
Just had a couple minor comments but this looks good to me!
